src/api_proxy/auth/auth_token.cc

```diff
@@ -15,188 +15,36 @@
-            GRPC_JSON_STRING);
-  FillChild(&json_type, &json_algorithm, &json_top, "typ", "JWT",
-            GRPC_JSON_STRING);
-
-  char *json_str = grpc_json_dump_to_string(&json_top, 0);
-  char *result = grpc_base64_encode(json_str, strlen(json_str), 1, 0);
-  gpr_free(json_str);
-  return result;
-}
-
-char *GenerateJwtClaim(const char *issuer, const char *subject,
-                       const char *audience, gpr_timespec token_lifetime) {
-  grpc_json json_top;
-  memset(&json_top, 0, sizeof(json_top));
-  json_top.type = GRPC_JSON_OBJECT;
-
-  gpr_timespec now = gpr_now(GPR_CLOCK_REALTIME);
-  gpr_timespec expiration = gpr_time_add(now, token_lifetime);
-  char now_str[GPR_LTOA_MIN_BUFSIZE];
-  char expiration_str[GPR_LTOA_MIN_BUFSIZE];
-  gpr_ltoa(now.tv_sec, now_str);
-  gpr_ltoa(expiration.tv_sec, expiration_str);
-
-  grpc_json json_issuer, json_subject, json_audience, json_now, json_expiration;
-  FillChild(&json_issuer, nullptr, &json_top, "iss", issuer, GRPC_JSON_STRING);
-  FillChild(&json_subject, &json_issuer, &json_top, "sub", subject,
-            GRPC_JSON_STRING);
-  FillChild(&json_audience, &json_subject, &json_top, "aud", audience,
-            GRPC_JSON_STRING);
-  FillChild(&json_now, &json_audience, &json_top, "iat", now_str,
-            GRPC_JSON_NUMBER);
-  FillChild(&json_expiration, &json_now, &json_top, "exp", expiration_str,
-            GRPC_JSON_NUMBER);
-
-  char *json_str = grpc_json_dump_to_string(&json_top, 0);
-  char *result = grpc_base64_encode(json_str, strlen(json_str), 1, 0);
-  gpr_free(json_str);
-  return result;
-}
+absl::optional<std::string> get_auth_token(const std::string& json_secret,
+                                           const std::string& audience) {
+  grpc_auth_json_key json_key =
+      grpc_auth_json_key_create_from_string(json_secret.c_str());
 
-char *GenerateSignatueHs256(const char *data, const char *key) {
-  grpc_slice key_buffer = grpc_base64_decode(key, 1);
-  if (GRPC_SLICE_IS_EMPTY(key_buffer)) {
-    gpr_log(GPR_ERROR, "Unable to decode base64 of secret");
-    return nullptr;
+  if (grpc_auth_json_key_is_valid(&json_key) == 0) {
+    return absl::nullopt;
   }
 
-  unsigned char res[256 / 8];
-  unsigned int res_len = 0;
-  HMAC(EVP_sha256(), GRPC_SLICE_START_PTR(key_buffer),
-       GRPC_SLICE_LENGTH(key_buffer),
-       reinterpret_cast<const unsigned char *>(data), strlen(data), res,
-       &res_len);
-  grpc_slice_unref(key_buffer);
-  if (res_len == 0) {
-    gpr_log(GPR_ERROR, "Cannot compute HMAC from secret.");
-    return nullptr;
-  }
-  return grpc_base64_encode(res, res_len, 1, 0);
-}
+  const char* token = grpc_jwt_encode_and_sign(&json_key, audience.c_str(),
+                                         TOKEN_LIFETIME, nullptr);
+  grpc_auth_json_key_destruct(&json_key);
 
-string DotConcat(const string &str1, const string &str2) {
-  return str1 + "." + str2;
+  return absl::optional<std::string>{token};
```

Comment:
I think you need to free token memory

Reply:
Good catch. Also confirmed the ASAN presubmit caught this. Done